impeller/typographer/backends/skia/typographer_context_skia.cc

```diff
@@ -244,6 +245,68 @@ static void DrawGlyph(SkCanvas* canvas,
                      sk_font,                                 // font
                      glyph_paint                              // paint
   );
+  canvas->restore();
+}
+
+/// @brief Batch render to a single surface.
+///
+/// This is only safe for use when updating a fresh texture.
+static bool BulkUpdateAtlasBitmap(const GlyphAtlas& atlas,
+                                  std::shared_ptr<BlitPass>& blit_pass,
+                                  HostBuffer& host_buffer,
+                                  const std::shared_ptr<Texture>& texture,
+                                  const std::vector<FontGlyphPair>& new_pairs,
+                                  size_t start_index,
+                                  size_t end_index) {
+  TRACE_EVENT0("impeller", __FUNCTION__);
+
+  bool has_color = atlas.GetType() == GlyphAtlas::Type::kColorBitmap;
+
+  SkBitmap bitmap;
+  bitmap.setInfo(GetImageInfo(atlas, Size(texture->GetSize())));
+  if (!bitmap.tryAllocPixels()) {
+    return false;
+  }
+
+  auto surface = SkSurfaces::WrapPixels(bitmap.pixmap());
+  if (!surface) {
+    return false;
+  }
+  auto canvas = surface->getCanvas();
+  if (!canvas) {
+    return false;
+  }
+
+  for (size_t i = start_index; i < end_index; i++) {
+    const FontGlyphPair& pair = new_pairs[i];
+    auto data = atlas.FindFontGlyphBounds(pair);
+    if (!data.has_value()) {
+      continue;
+    }
+    auto [pos, bounds] = data.value();
+    Size size = pos.GetSize();
+    if (size.IsEmpty()) {
+      continue;
+    }
+
+    DrawGlyph(canvas, SkPoint::Make(pos.GetLeft(), pos.GetTop()),
+              pair.scaled_font, pair.glyph, bounds, pair.glyph.properties,
+              has_color);
+  }
+
+  // Writing to a malloc'd buffer and then copying to the staging buffers
+  // benchmarks as substantially faster on a number of Android devices.
+  BufferView buffer_view = host_buffer.Emplace(
```

Comment:
You could reserve an arena in the host buffer itself and wrap that arena as an SkBitmap to reduce one copy. There is a HostBuffer::Emplace overload that takes a callback that will give you the pointer to the buffer.

Reply:
If you recall from a while back: Skia is going to read from the buffer (even with BlendMode.src rendering) while rendering the glyphs, and the device buffers are not host read cached - so the performance tanks

Reply:
Ah yeah, I remember now. Add a comment to that effect maybe?